octoprint_firmwareupdate/__init__.py

```diff
@@ -101,6 +107,32 @@ def on_after_startup(self):
         else:
             self._logger.info("Auto firmware update disabled, skipping...")
 
+    @octoprint.plugin.BlueprintPlugin.route("/upload", methods=["POST"])
+    @restricted_access
+    @admin_permission.require(403)
+    def upload_file(self):
+        # Delete any firmware files that may exist when using custom firmware
+        self._delete_firmware_files()
+        if "base64String" not in flask.request.values:
+            return flask.make_response("Expected a base64String value", 400)
+
+        try:
+            decode = base64.b64decode(flask.request.values['base64String'])
+            self._check_directories()
+            with open(os.path.join(self.firmware_directory,
+                                   "firmware.hex"), "wb") as firmware:
+                firmware.write(decode)
+                firmware.close()
+        except (TypeError, IOError):
+            error_text = "There was an issue saving the firmware file."
+            self._logger.warn(error_text)
+            self._update_status(
+                False, "error", error_text)
+            return flask.make_response(error_text, 400)
```

Comment:
I would log or do something with the original exception.  If you ever find yourself here, and you need to debug it, you need to know the original error message and error type.